### PR TITLE
feat: commands download user files at startup so k8s can support larger context directories [DET-8830]

### PR DESCRIPTION
**Pull request by NicholasBlaskey**
## Description

We hacked around k8s using config maps to pass user files for experiments by downloading the user provided files at startup. 

We will also use this same hack for command files and make downloading a model def based off task ID.

## Test Plan

Int + e2e tests

Manual

(Do this on a Kubernetes Det cluster) 

Create a few "large" files
```
for i in $(seq 3) ; do cat /dev/urandom| base64 | head -c 1000000 > 1mb.$i ; done
```

Run a notebook
```
det notebook start --context .
```

Make sure you can see the 1mb files in the notebook in the webui.

Also make sure we didn't bork context directory stuff.

From determined repo root run
```
det cmd run -c examples/ "ls"
```

We should see
```
[2023-09-14T15:32:52.198947Z]          || INFO: Scheduling Command (exactly-huge-mongoose) (id: 1144863e-243b-484c-8340-1e278462e24a.1)
[2023-09-14T15:32:52.719153Z]          || INFO: Command (exactly-huge-mongoose) was assigned to an agent
[2023-09-14T15:32:52.000000Z] 1dc7ea1b || INFO: Pod cmd-1144863e-243b-484c-8340-1e278462e24a-0-1144863e-243b-484c-8340-1e278462e24a.1-relevant-husky: Pod resources allocated.
[2023-09-14T15:32:53.000000Z] 1dc7ea1b || INFO: Pod cmd-1144863e-243b-484c-8340-1e278462e24a-0-1144863e-243b-484c-8340-1e278462e24a.1-relevant-husky: Container image "determinedai/environments:cuda-11.3-pytorch-1.12-tf-2.11-gpu-2b7e2a1" already present on machine
[2023-09-14T15:32:53.000000Z] 1dc7ea1b || INFO: Pod cmd-1144863e-243b-484c-8340-1e278462e24a-0-1144863e-243b-484c-8340-1e278462e24a.1-relevant-husky: Created container determined-init-container
[2023-09-14T15:32:53.000000Z] 1dc7ea1b || INFO: Pod cmd-1144863e-243b-484c-8340-1e278462e24a-0-1144863e-243b-484c-8340-1e278462e24a.1-relevant-husky: Started container determined-init-container
[2023-09-14T15:32:53.000000Z] 1dc7ea1b || INFO: Pod cmd-1144863e-243b-484c-8340-1e278462e24a-0-1144863e-243b-484c-8340-1e278462e24a.1-relevant-husky: Container image "determinedai/environments:cuda-11.3-pytorch-1.12-tf-2.11-gpu-2b7e2a1" already present on machine
[2023-09-14T15:32:53.000000Z] 1dc7ea1b || INFO: Pod cmd-1144863e-243b-484c-8340-1e278462e24a-0-1144863e-243b-484c-8340-1e278462e24a.1-relevant-husky: Created container determined-container
[2023-09-14T15:32:53.000000Z] 1dc7ea1b || INFO: Pod cmd-1144863e-243b-484c-8340-1e278462e24a-0-1144863e-243b-484c-8340-1e278462e24a.1-relevant-husky: Started container determined-container
[2023-09-14T15:32:54.662758Z] 1dc7ea1b || INFO: Resources for Command (exactly-huge-mongoose) have started
[2023-09-14T15:32:58.658894Z] 1dc7ea1b || Makefile
[2023-09-14T15:32:58.658973Z] 1dc7ea1b || README.md
[2023-09-14T15:32:58.659005Z] 1dc7ea1b || build
[2023-09-14T15:32:58.659062Z] 1dc7ea1b || computer_vision
[2023-09-14T15:32:58.659135Z] 1dc7ea1b || custom_search_method
[2023-09-14T15:32:58.659196Z] 1dc7ea1b || deepspeed
[2023-09-14T15:32:58.659349Z] 1dc7ea1b || deepspeed_autotune
[2023-09-14T15:32:58.659392Z] 1dc7ea1b || diffusion
[2023-09-14T15:32:58.659433Z] 1dc7ea1b || features
[2023-09-14T15:32:58.659457Z] 1dc7ea1b || gan
[2023-09-14T15:32:58.659552Z] 1dc7ea1b || graphs
[2023-09-14T15:32:58.659626Z] 1dc7ea1b || hf_trainer_api
[2023-09-14T15:32:58.659658Z] 1dc7ea1b || hp_search_benchmarks
[2023-09-14T15:32:58.659679Z] 1dc7ea1b || meta_learning
[2023-09-14T15:32:58.659771Z] 1dc7ea1b || nas
[2023-09-14T15:32:58.659845Z] 1dc7ea1b || nlp
[2023-09-14T15:32:58.659932Z] 1dc7ea1b || tests
[2023-09-14T15:32:58.660009Z] 1dc7ea1b || tutorials
[2023-09-14T15:32:59.682209Z] 1dc7ea1b || INFO: resources exited successfully with a zero exit code
[2023-09-14T15:32:59.688139Z]          || INFO: Command (exactly-huge-mongoose) was terminated: allocation stopped early after all resources exited: resources exited successfully with a zero exit code
```

Running 
```
det cmd run -i examples/ "ls ./examples"
```

we should see the same 

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/api_command.go
Comment (stoksc) [inline master/internal/api_command.go:188]: can we call this `contextDirectory` or something, instead of model def? here and elsewhere,

File: e2e_tests/tests/command/test_run.py
Comment (stoksc) [inline e2e_tests/tests/command/test_run.py:205]: may as well just test this with integrations, honestly. i dont see that this test adds a lot if the APIs are tested and the experiment test already exists,

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:660]: would be nice if we persisted and evicted this much earlier.

File: master/pkg/model/task.go
Comment (stoksc) [inline master/pkg/model/task.go:415]: personally, i would name this `task_context_directory` all over, store it in a table called that, with `(task_id, context_directory)` and eventually move trials there.

File: master/pkg/tasks/task_command.go
Comment (stoksc) [inline master/pkg/tasks/task_command.go:52]: this is only for command restore or something, right? and it never uses it, so we dont need line 103, we just need it here to not explode when pulling it from the db?

**Review by stoksc — COMMENTED** [review]
very nice, minor comments, mostly about naming.

File: master/pkg/model/task.go
Comment (NicholasBlaskey) [reply master/pkg/model/task.go:415]: I think there is something a little awkward with this approach, since experiments can have multiple tasks but they will always have the same context directory.

Also I don't know if we could ever pay that migration cost.

Is ```ntsc_context_directory``` better? I don't wanna add more confusion to the question of what a task is.

File: master/internal/command/command.go
Comment (NicholasBlaskey) [reply master/internal/command/command.go:660]: how much earlier? 

Do you mean like in the api calls?

Comment (stoksc) [reply master/internal/command/command.go:660]: yeah, but i dont think it is feasible for this PR

File: master/pkg/model/task.go
Comment (stoksc) [reply master/pkg/model/task.go:415]: put it on the job instead? though that is technically not correct because checkpoint gc runs different code.

Comment (ioga) [reply master/pkg/model/task.go:415]: I am not sure we should be overloading jobs beyond moving this atomic thing up/down in the queue or between resource pools.

e.g. looking to the future nested tasks proposal, entire task tree should probably be the same job, but the individual tasks within a tree could be running different code.

Comment (stoksc) [reply master/pkg/model/task.go:415]: so you agree with `task_context_directory`?

Comment (stoksc) [reply master/pkg/model/task.go:415]: i think, much to your dismay, the job concept is already getting overloaded.

File: e2e_tests/tests/command/test_run.py
Comment (NicholasBlaskey) [reply e2e_tests/tests/command/test_run.py:205]: I do like this test, (I'll remove the other part from e2e_cpu and just have it run in e2e_k8s) as a smoke test

We have had a bunch of people report this issue so I think it makes sense to know we support it forever with this user perspective test. I guess its really not likely it will fail based on the how system works.

Comment (NicholasBlaskey) [reply e2e_tests/tests/command/test_run.py:205]: What integration tests are you thinking to add? 

could probaly test command's prestart but that is maybe only slightly annoying with actors.

File: master/pkg/tasks/task_command.go
Comment (NicholasBlaskey) [reply master/pkg/tasks/task_command.go:52]: 103 as in right? https://github.com/determined-ai/determined/blob/3f9a980ec1eec651f60d07b69327adebf181e8e1/master/pkg/tasks/task_command.go#L103

Yeah I think this is where it will be eventually pulled from the database
https://github.com/determined-ai/determined/blob/3f9a980ec1eec651f60d07b69327adebf181e8e1/master/internal/command/models.go#L24

File: master/pkg/model/task.go
Comment (NicholasBlaskey) [reply master/pkg/model/task.go:415]: I think its pretty clear ```task_id, context_directory``` will be the structure of the table

I'll call it ```task_context_directory``` for now, we can always rename it if we build on it 

and I will add a comment explaining why it doesn't have trials

File: e2e_tests/tests/command/test_run.py
Comment (NicholasBlaskey) [reply e2e_tests/tests/command/test_run.py:205]: Added a command integration test, making the large context test only run on k8s tests (once per commit merged) rather than on every e2e_test

**Comment by NicholasBlaskey** [comment]
@EmilyBonar the proto changes shouldn't touch anything web, so should be an easy review

File: docs/release-notes/k8s-context-size-limit.rst
Comment (tara-det-ai) [inline docs/release-notes/k8s-context-size-limit.rst:7]: :orphan:

**Improvements**

-  Kubernetes: Increased the file context limit for Notebooks, Commands, Tensorboards, and Shells in Kubernetes from approximately 1MB to roughly 95MB, the same limit as the agent resource manager.

**Review by tara-det-ai — COMMENTED** [review]
suggested edit

**Review by EmilyBonar — APPROVED** [review]
Yeah, looks like the web changes are just a few new (unused) generated types/functions. LGTM as long as the backend implementations are solid.

File: harness/determined/exec/prep_container.py
Comment (wes-turner) [inline harness/determined/exec/prep_container.py:342]: If `trial` is deprecated, could you please add a [FutureWarning](https://docs.python.org/3/library/exceptions.html#FutureWarning) if it's used?

If `trial` is not deprecated, then does it make sense to not introduce `download_context_directory` and just continue using `trial`?

File: harness/determined/exec/prep_container.py
Comment (wes-turner) [inline harness/determined/exec/prep_container.py:304]: You're following the example set by existing arguments, but in the spirit of continuous improvement, could you please give this a short [help](https://docs.python.org/3/library/argparse.html#help)?

File: harness/determined/exec/prep_container.py
Comment (wes-turner) [inline harness/determined/exec/prep_container.py:62]: nit: It's not a bug or anything, but this seems more complicated than
```python
b64_tgz = context_directory_resp.b64Tgz
```
for the same functionality.

File: harness/determined/exec/prep_container.py
Comment (wes-turner) [inline harness/determined/exec/prep_container.py:24]: Could this info be passed with the `info` parameter instead of read fresh from the environment? It always makes me uncomfortable to see globals injected into code -- especially when they're not typed.

Comment (wes-turner) [reply harness/determined/exec/prep_container.py:24]: I think what this comes down to is that some parameters are passed via flags, and some via environment variables, and it's a little hard to track as a reader. There's only so much that you can do about that here right now, but I think that this is a request to read/parse the various parameters in one place. Just like `download_context_directory` isn't the right place to parse flags, it's also probably not the right place to parse parameters read from the environment.

Comment (NicholasBlaskey) [reply harness/determined/exec/prep_container.py:24]: I moved the "```trial_info._from_env``` to ```trial_info._to_file```" parsing to where we originally parse ```info```

I was also doing a double parse of ```ClusterInfo```, so this seems a lot better

File: harness/determined/exec/prep_container.py
Comment (wes-turner) [reply harness/determined/exec/prep_container.py]: Looks great!

File: harness/determined/exec/prep_container.py
Comment (wes-turner) [reply harness/determined/exec/prep_container.py:358]: > Does anyone call this script outside of our entrypoints or could it be considered "internal"?

I for sure don't have the context to *actually* know this, but I do think "internal" is a little nuanced. I _believe_ that this is called by (only by?) determined processes from code that, strictly speaking, isn't a part of the `determined-ai/determined` repo. Or at the very least, code that's not a part of `determined-ai/determined/harness`. So in that sense it's at least got a caller not internal to the package it's a part of.

But that's probably a pedantic distinction, and you're asking if we can just remove the flag altogether, no worries about the deprecation dance. I don't even know, here, which team to ask.

Anyway, following the deprecation path, this looks great. Only outstanding suggestion is to remove the `stacklevel` argument, since `prep_container.py` is being called as an executable and it looks funny to assign the warning to Python internals.

File: harness/determined/exec/prep_container.py
Comment (wes-turner) [reply harness/determined/exec/prep_container.py]: Nice. I think this looks really good, too.

**Review by wes-turner — APPROVED** [review]
Looks good!

File: docs/release-notes/k8s-context-size-limit.rst
Comment (stoksc) [inline docs/release-notes/k8s-context-size-limit.rst:6]: redundant "in Kubernetes"

File: e2e_tests/tests/command/test_run.py
Comment (stoksc) [inline e2e_tests/tests/command/test_run.py:205]: ticket?

File: master/internal/api_command.go
Comment (stoksc) [inline master/internal/api_command.go:176]: offline chat: bikeshedding about function size

File: master/internal/api_command.go
Comment (stoksc) [inline master/internal/api_command.go:364]: I wrote the below comment before I read back up and saw `launchReq.Spec` was a pointer, but I'm gonna post it anyway as some hand wavy argument that this shallow is confusing / error-prone, anyway.

does this work? pretty sure this shallow copy is going to make some of the later mutates ineffective (e.g. `spec.Config.Description, see https://go.dev/play/p/sa4sIHVOELg)

File: master/internal/api_experiment.go
Comment (stoksc) [inline master/internal/api_experiment.go:2105]: 10/10 name

File: master/internal/api_experiment.go
Comment (stoksc) [inline master/internal/api_experiment.go:2108]: nit context in the error; presumably the caller has enough context to know which task ID it was. i've made a call to not include identical info in other calls for this reason, wdyt?

**Review by stoksc — APPROVED** [review]
lgtm

File: master/internal/api_experiment.go
Comment (NicholasBlaskey) [reply master/internal/api_experiment.go:2108]: As support for including it K8s API will includes it
```
kubectl get node thisnodedontexist
Error from server (NotFound): nodes "thisnodedontexist" not found
```

I'm not sure exactly the reasons why Kubernetes does this or if an API style guide says one or the other. 

I think I mostly have been doing it since it is frustating to call and API and think you have the correct ID but the server tells you not found and you are not sure why.

Comment (stoksc) [reply master/internal/api_experiment.go:2108]: I can buy that. Gotta be careful to add that context only once, though, so you don't end up with "doing thing for abc: load part of abc: xyz abc.."

**Pull request merged**
